src/tools/illink/src/linker/Linker.Steps/MarkScopeStack.cs

```diff
@@ -51,7 +51,7 @@ public void Dispose ()
 			}
 		}
 
-		readonly struct ParentScope : IDisposable
+		internal readonly struct ParentScope : IDisposable
```

Comment:
I would be careful here -- the contract for `IDisposable` is that `Dispose` will be called *at least once*. If it's called multiple times, it looks like this code might be broken.

One way to avoid that is to make this struct mutable and either make `scopeStack` nullable and null it out when Dispose is called, or keep an extra flag for disposed.

Lastly, mutable structs are dangerous in that they can be easily copied by boxing and other things. There's no real way to fix that right now (C# doesn't have a way to prevent it), it just means you need to be very careful and try to keep the struct as private as possible.

Reply:
I think the checks that throw `InternalErrorException` should handle the case where `Dispose` is called twice. We could make it more resilient with a check that doesn't do anything on a second `Dispose`, but I don't think we've ever hit an issue with it before.

Reply:
Got it -- the intent is to throw on multiple dispose. That makes sense, although that is not the contract that's stated on IDisposable -- it says that it's OK to call Dispose multiple times. But I think having stricter behavior is OK here.